### PR TITLE
Add rmc-compiler binary to drive rust compilation

### LINKED ISSUE #42: Compiler interface: remove need to patch upstream API

### PR DESCRIPTION
**Pull request by celinval**
### Description of changes: 

I have created a new compilation driver (rmc-compiler) that no longer depends on a modified version of the rust compiler, neither depends on the bootstrap script. 

### Resolved issues:

Resolves #42 and #658

### Call-outs:

I'm not quite happy with a few things in this change, but it is getting way too big to manage in a separate branch. It is also getting quite big to be reviewed. Thus, I created the following issues to track a few improvements:

- #719: Improve the layout of our packages and create a new workspace.
- #720: Improve how we are building our libraries. For now, the rmc-compiler has a build-dependency that ensures the libraries get built with the same toolchain as rmc-compiler.
- #721: Improve the interface between cargo rmc and the rmc-compiler. The solution here is a bit hacky. There are a few tricky things here. Cargo invoke rustc to retrieve its version as well as to compile build scripts which shouldn't use the goto-c backend.

### Testing:

* How is this change tested? All our tests.

* Is this a refactor change? Yes

### Checklist
- [ ] Each commit message has a non-empty body, explaining why the change was made
- [ ] Methods or procedures are documented
- [ ] Regression or unit tests are included, or existing tests cover the modified code
- [ ] My PR is restricted to a single feature or bugfix

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 and MIT licenses.

### CODE REVIEW

File: src/rmc-compiler/build.rs
Comment (tedinski) [inline src/rmc-compiler/build.rs:14]: Just checking in here, because I'll have to work on the install later, that you're dynamic linking against a `rustup` provided rust compiler library?

Is there documentation on how this works somewhere? (e.g. I'm used to rust's normal static linking, so dynamic linking to a rust library is not something I'm familiar with)

Also, this might complication installation later. For one, we need the correct relative RPATH here not an absolute path like it is now, but also we'll need to distribute the correct rustc version, right? Or is there a stable-ish ABI here?

If there's not documentation about this, I think we need to write one.

File: src/rmc-compiler/src/main.rs
Comment (tedinski) [inline src/rmc-compiler/src/main.rs:54]: Do you have a preference for `clap`? I was starting work on cargo-rmc with `structopt` which is a wrapper around `clap`. Just curious if you think I shouldn't be using that for some reason (it's not too late for me to abort and turn around...)

File: src/rmc-compiler/src/main.rs
Comment (tedinski) [inline src/rmc-compiler/src/main.rs:25]: Do you have any objection to my moving this into `cargo-rmc` in the future instead of having a `--goto-c` flag to turn them on? That was my original plan. (This is fine for now ofc)

File: src/rmc-compiler/src/main.rs
Comment (tedinski) [inline src/rmc-compiler/src/main.rs:183]: are these environment variables typically provided by cargo? Or is this just another option we're creating to supply them?

File: scripts/rmc-rustc
Comment (tedinski) [inline scripts/rmc-rustc:46]: what is `${RMCFLAGS}` here. I mean, I can search this diff, but there should be a comment saying what's happening here.

File: library/rmc_macros/src/lib.rs
Comment (celinval) [reply library/rmc_macros/src/lib.rs:37]: No. This is actually the only I could find to turn on and off this configuration via cargo. I thought the plan was for users to be able to use cargo here.

File: src/rmc-compiler/build.rs
Comment (celinval) [reply src/rmc-compiler/build.rs:14]: Yes, you are correct. It is linking to a rustup shared library.  We will need to make it relative if we want to distribute the binaries or if we eventually install it via rustup.

This should work if the user wants to build / install it with cargo.

File: src/rmc-compiler/src/main.rs
Comment (celinval) [reply src/rmc-compiler/src/main.rs:54]: I don't have any strong preference. I haven't done much research since this code is throw away code. Whatever works.

File: src/rmc-compiler/src/main.rs
Comment (celinval) [reply src/rmc-compiler/src/main.rs:25]: Let's talk more about this one. I'm not 100% sure what the solution that you are proposing is.

File: src/rmc-compiler/src/main.rs
Comment (celinval) [reply src/rmc-compiler/src/main.rs:183]: Yes, they are usually provided by cargo. This is similar to code from clippy and Miri.

File: scripts/rmc-rustc
Comment (celinval) [reply scripts/rmc-rustc:46]: Sure

Comment (celinval) [reply scripts/rmc-rustc:46]: It's set by rmc.py because cargo does some funny stuff with RUSTFLAGS.

**Comment by tedinski** [comment]
Also, if I'm not mistaken, if we eliminate our remaining use of `x.py` (just compiletest and dashboard/bookrunner, right?) CI runtimes are about to be _outstanding_!

File: library/rmc_macros/src/lib.rs
Comment (tedinski) [reply library/rmc_macros/src/lib.rs:37]: So users should be putting `[dev-dependencies] rmc = "1.0"` or whatever into their Cargo.toml, but that means when cargo-rmc does its build/codegen of the project `cfg(rmc)` would be set during that build.

So I was suggesting the current approach of building this library at the same time as `rmc-compiler` is a temporary hack (to ensure it's already built to support the current `rmc` script).

File: src/rmc-compiler/build.rs
Comment (tedinski) [reply src/rmc-compiler/build.rs:14]: Do you know where docs about this are? I searched "rust shared library rustc_driver" and I'm not finding anything helpful. I'm not sure how fragile this approach is. (But I sure like what it does for CI times! Just trying to understand the implications for installation later...)

File: library/rmc_macros/src/lib.rs
Comment (celinval) [reply library/rmc_macros/src/lib.rs:37]: That was a suggestion from Daniel to pre-build the rmc library. I can pre-build with the rmc config enable using RUSTFLAGS then. Let me try this and update the PR.

File: src/rmc-compiler/build.rs
Comment (celinval) [reply src/rmc-compiler/build.rs:14]: I couldn't find any doc either. I did a lot of reverse engineering in other tools that use rustc to figure this out and I also talked to a few folks.

This seems to be the de facto way to build tools on the top of rustc. See https://github.com/rust-lang/rust-clippy/blob/830f2205d4002b26f7fd36d8467fb491eb2306ac/src/driver.rs, https://github.com/rust-lang/miri/blob/035933186957cf81c488261fb48a98bf523e8006/src/lib.rs and this little example: https://doc.rust-lang.org/beta/unstable-book/language-features/plugin.html?highlight=rustc_private#lint-plugins

**Comment by celinval** [comment]
> Also, if I'm not mistaken, if we eliminate our remaining use of `x.py` (just compiletest and dashboard/bookrunner, right?) CI runtimes are about to be _outstanding_!

For sure! This change reduced the CI in about 10min. @adpaco-aws and I are working on the second piece. After that, we should be able to prune our repository, remove the bootstrap script and no longer depend on the rust upstream. :)

File: src/rmc-compiler/rust-toolchain.toml
Comment (tedinski) [inline src/rmc-compiler/rust-toolchain.toml:5]: Does this cause cargo to download this toolchain, or does it fail or something if it's not present?

Comment (celinval) [reply src/rmc-compiler/rust-toolchain.toml:5]: It will download it.

Comment (celinval) [reply src/rmc-compiler/rust-toolchain.toml:5]: I believe any rustup command will trigger the download

File: src/rmc-compiler/src/main.rs
Comment (tedinski) [inline src/rmc-compiler/src/main.rs:101]: ah, I see, so here's where we have the rustc flags after our flags, and we have the problem with cargo mixing them up. And that's why we need `RMCFLAGS` in `rmc-rustc` for now.

**Review by tedinski — APPROVED** [review]
LGTM. Think DSN should review too?

**Comment by celinval** [comment]
> LGTM. Think DSN should review too?

I think it would be nice to have someone else reviewing it. Let me ping @danielsn.

File: src/rmc-compiler/src/main.rs
Comment (danielsn) [inline src/rmc-compiler/src/main.rs:133]: Will this give enough debug info?

File: src/rmc-compiler/src/main.rs
Comment (danielsn) [inline src/rmc-compiler/src/main.rs:137]: How will this be used once we do?

File: src/rmc-compiler/src/main.rs
Comment (danielsn) [inline src/rmc-compiler/src/main.rs:164]: Is `to_string_lossy` what we want here? If there are non-valid characters, shouldn't we just fail?

File: src/rmc-compiler/README.md
Comment (danielsn) [inline src/rmc-compiler/README.md:7]: built

**Review by danielsn — APPROVED** [review]
Some questions, but LGTM

File: scripts/cargo-rmc
Comment (celinval) [reply scripts/cargo-rmc:29]: This was just to make it easier to debug codegen... I was also thinking about using that in our tests, since today we call rustc-rmc directly.

File: src/rmc-compiler/cbmc/src/lib.rs
Comment (celinval) [reply src/rmc-compiler/cbmc/src/lib.rs:39]: I agree. We should create a jira for that.

File: src/rmc-compiler/rustc_codegen_rmc/src/lib.rs
Comment (celinval) [reply src/rmc-compiler/rustc_codegen_rmc/src/lib.rs:36]: Yes. Unfortunately it is.

File: src/rmc-compiler/rmc_queries/src/lib.rs
Comment (celinval) [reply src/rmc-compiler/rmc_queries/src/lib.rs:7]: Yes. I don't have to add them now. But I wanted to start trying the concept sooner rather than later.

File: src/rmc-compiler/rustc_codegen_rmc/src/context/goto_ctx.rs
Comment (celinval) [reply src/rmc-compiler/rustc_codegen_rmc/src/context/goto_ctx.rs:62]: That's a good point. I could provide the query here.

File: src/rmc-compiler/rustc_codegen_rmc/src/compiler_interface.rs
Comment (celinval) [reply src/rmc-compiler/rustc_codegen_rmc/src/compiler_interface.rs:65]: Is this related to one of your next comments that we should pass the queries instead, right? If so, I like the suggestion.

File: src/rmc-compiler/rustc_codegen_rmc/src/context/goto_ctx.rs
Comment (celinval) [reply src/rmc-compiler/rustc_codegen_rmc/src/context/goto_ctx.rs:66]: I didn't dive deep here, but this variable is passed to the constructor of VtableCtx.

File: src/rmc-compiler/src/main.rs
Comment (celinval) [reply src/rmc-compiler/src/main.rs:133]: This won't suppress the output of rustc. The error should've been printed already.

File: src/rmc-compiler/src/main.rs
Comment (celinval) [reply src/rmc-compiler/src/main.rs:128]: If it's not present, we will invoke `rustc` compilation without `RMC` arguments. This is required due to our current implementation of cargo rmc. Cargo calls rustc to query its version as well as to build `build.rs` scripts.

Comment (celinval) [reply src/rmc-compiler/src/main.rs:128]: So we need to support both options, with and without --goto-c.

File: src/rmc-compiler/src/main.rs
Comment (celinval) [reply src/rmc-compiler/src/main.rs:137]: Callbacks can be used to analyze each compilation step and it can also be used to stop compilation in earlier stages. For example, if we want to traverse the AST after parsing, we can implement a callback.

File: src/rmc-compiler/src/main.rs
Comment (celinval) [reply src/rmc-compiler/src/main.rs:164]: Sure. We can fail.

File: src/rmc-compiler/rustc_codegen_rmc/src/context/goto_ctx.rs
Comment (danielsn) [reply src/rmc-compiler/rustc_codegen_rmc/src/context/goto_ctx.rs:66]: But I think we don't need to assign it to itself first

File: src/rmc-compiler/rust-toolchain.toml
Comment (celinval) [reply src/rmc-compiler/rust-toolchain.toml:5]: The plan is to update this every two weeks for now. Instead of merging with the upstream repository, we will keep up with the latest compiler by updating the toolchain file.

I don't know long they plan to keep old toolchains, but from what I've seen, they have been keeping them all on S3 and some linux distro uses old versions of rustup. For documentation purpose, we could also track the commit that is associated with the current version.

Unfortunately, we still rely on a lot of nightly features, so we can't just use stable releases.

File: src/rmc-compiler/cbmc/src/lib.rs
Comment (celinval) [reply src/rmc-compiler/cbmc/src/lib.rs:27]: It allow us to use rustc crates.

File: src/rmc-compiler/rustc_codegen_rmc/src/context/goto_ctx.rs
Comment (celinval) [reply src/rmc-compiler/rustc_codegen_rmc/src/context/goto_ctx.rs:66]: ah, I see. Yeah, this is garbage. :)

File: src/rmc-compiler/build.rs
Comment (danielsn) [inline src/rmc-compiler/build.rs:10]: Did we want to do a lossy conversion, or to fail fast here?

File: src/rmc-compiler/build.rs
Comment (danielsn) [inline src/rmc-compiler/build.rs:16]: why `..` twice? Should one be `.`

**Review by danielsn — CHANGES_REQUESTED** [review]
I think there is an issue where you meant to say `.` and write `..` for a directory path.  Not sure, but blocking merge till we look at it

File: src/rmc-compiler/build.rs
Comment (celinval) [reply src/rmc-compiler/build.rs:16]: Because this is inside src/rmc-compiler/ so we need to walk back two directories.

File: src/rmc-compiler/build.rs
Comment (celinval) [reply src/rmc-compiler/build.rs:10]: Sure, I can fall fast.

**Review by danielsn — APPROVED** [review]

File: src/rmc-compiler/build.rs
Comment (danielsn) [reply src/rmc-compiler/build.rs:16]: Ahh, I thought this was a search path, not a set of folders

**Pull request merged**
